File: modules/util/color_test.go
// Copyright 2023 The Gitea Authors. All rights reserved.
// SPDX-License-Identifier: MIT
package util

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func Test_HexToRBGColor(t *testing.T) {
	cases := []struct {
		colorString string
		expectedR   float64
		expectedG   float64
		expectedB   float64
	}{
		{"2b8685", 43, 134, 133},
		{"1e1", 17, 238, 17},
		{"#1e1", 17, 238, 17},
		{"1e16", 17, 238, 17},
		{"3bb6b3", 59, 182, 179},
		{"#3bb6b399", 59, 182, 179},
		{"#0", 0, 0, 0},
		{"#00000", 0, 0, 0},
		{"#1234567", 0, 0, 0},
	}
	for n, c := range cases {
		r, g, b := HexToRBGColor(c.colorString)
		assert.InDelta(t, c.expectedR, r, 0, "case %d: error R should match: expected %f, but get %f", n, c.expectedR, r)
		assert.InDelta(t, c.expectedG, g, 0, "case %d: error G should match: expected %f, but get %f", n, c.expectedG, g)
		assert.InDelta(t, c.expectedB, b, 0, "case %d: error B should match: expected %f, but get %f", n, c.expectedB, b)
	}
}

func Test_UseLightText(t *testing.T) {
	cases := []struct {
		color    string
		expected string
	}{
		{"#d73a4a", "#fff"},
		{"#0075ca", "#fff"},
		{"#cfd3d7", "#000"},
		{"#a2eeef", "#000"},
		{"#7057ff", "#fff"},
		{"#008672", "#fff"},
		{"#e4e669", "#000"},
		{"#d876e3", "#000"},
		{"#ffffff", "#000"},
		{"#2b8684", "#fff"},
		{"#2b8786", "#fff"},
		{"#2c8786", "#000"},
		{"#3bb6b3", "#000"},
		{"#7c7268", "#fff"},
		{"#7e716c", "#fff"},
		{"#81706d", "#fff"},
		{"#807070", "#fff"},
		{"#84b6eb", "#000"},
	}
	for n, c := range cases {
		assert.Equal(t, c.expected, ContrastColor(c.color), "case %d: error should match", n)
	}
}

File: web_src/js/utils/color.test.ts
import {contrastColor} from './color.ts';

test('contrastColor', () => {
  expect(contrastColor('#d73a4a')).toBe('#fff');
  expect(contrastColor('#0075ca')).toBe('#fff');
  expect(contrastColor('#cfd3d7')).toBe('#000');
  expect(contrastColor('#a2eeef')).toBe('#000');
  expect(contrastColor('#7057ff')).toBe('#fff');
  expect(contrastColor('#008672')).toBe('#fff');
  expect(contrastColor('#e4e669')).toBe('#000');
  expect(contrastColor('#d876e3')).toBe('#000');
  expect(contrastColor('#ffffff')).toBe('#000');
  expect(contrastColor('#2b8684')).toBe('#fff');
  expect(contrastColor('#2b8786')).toBe('#fff');
  expect(contrastColor('#2c8786')).toBe('#000');
  expect(contrastColor('#3bb6b3')).toBe('#000');
  expect(contrastColor('#7c7268')).toBe('#fff');
  expect(contrastColor('#7e716c')).toBe('#fff');
  expect(contrastColor('#81706d')).toBe('#fff');
  expect(contrastColor('#807070')).toBe('#fff');
  expect(contrastColor('#84b6eb')).toBe('#000');
});
